File: pipe.go
package main

import (
	"fmt"
	"io"
	"time"
)

func main() {
	r, w := io.Pipe()
	go func() {
		n, err := w.Write([]byte("123"))
		if err != nil {
			fmt.Println("write error: ", err)
			return
		}
		fmt.Printf("write %d bytes\n", n)
	}()

	buf := make([]byte, 100)

	go func() {
		n, err := r.Read(buf)
		if err != nil {
			fmt.Println("read error: ", err)
			return
		}
		fmt.Printf("read %d bytes\n", n)
		fmt.Println(string(buf))
	}()

	time.Sleep(time.Second * 3)
}

File: pipe.c
#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <sys/wait.h>

int main() {
    int pipefd[2];
    pid_t cpid;

    if (pipe(pipefd) == -1) {
        perror("pipe");
        exit(EXIT_FAILURE);
    }

    cpid = fork();
    switch (cpid) {
    case -1:
        perror("fork");
        exit(EXIT_FAILURE);
    case 0: // child
        close(pipefd[0]);
        char *s = "123\n";
        write(pipefd[1], s, strlen(s));
        close(pipefd[1]);
        _exit(EXIT_SUCCESS);
    default:
        close(pipefd[1]);
        char buf[1024];
        read(pipefd[0], &buf, 1024);
        printf("%s\n", buf);
        close(pipefd[0]);
        wait(NULL);
        _exit(EXIT_SUCCESS);
    }
}
